packages/flutter/lib/src/widgets/scroll_position.dart

```diff
@@ -277,14 +301,83 @@ abstract class ScrollPosition extends ViewportOffset with ScrollMetrics {
   Drag drag(DragStartDetails details, VoidCallback dragCancelCallback);
 
   @protected
-  void didUpdateScrollPositionBy(double delta);
+  ScrollActivity get activity => _activity;
+  ScrollActivity _activity;
 
-  @protected
-  void didOverscrollBy(double value);
+  /// Change the current [activity], disposing of the old one and
+  /// sending scroll notifications as necessary.
+  ///
+  /// If the argument is null, this method has no effect. This is convenient for
+  /// cases where the new activity is obtained from another method, and that
+  /// method might return null, since it means the caller does not have to
+  /// explictly null-check the argument.
+  void beginActivity(ScrollActivity newActivity) {
+    if (newActivity == null)
+      return;
+    bool wasScrolling, oldIgnorePointer;
+    if (_activity != null) {
+      oldIgnorePointer = _activity.shouldIgnorePointer;
+      wasScrolling = _activity.isScrolling;
+      if (wasScrolling && !newActivity.isScrolling)
+        didEndScroll();
+      _activity.dispose();
+    } else {
+      oldIgnorePointer = false;
+      wasScrolling = false;
+    }
+    _activity = newActivity;
+    isScrollingNotifier.value = activity.isScrolling;
+    if (oldIgnorePointer != activity.shouldIgnorePointer)
+      context.setIgnorePointer(activity.shouldIgnorePointer);
+    if (!wasScrolling && _activity.isScrolling)
+      didStartScroll();
+  }
+
+
+  // NOTIFICATION DISPATCH
+
+  /// Called by [beginActivity] to report when an activity has started.
+  void didStartScroll() {
+    activity.dispatchScrollStartNotification(cloneMetrics(), context.notificationContext);
+  }
+
+  /// Called by [setPixels] to report a change to the [pixels] position.
+  void didUpdateScrollPositionBy(double delta) {
+    activity.dispatchScrollUpdateNotification(cloneMetrics(), context.notificationContext, delta);
+  }
+
+  /// Called by [beginActivity] to report when an activity has ended.
+  void didEndScroll() {
+    activity.dispatchScrollEndNotification(cloneMetrics(), context.notificationContext);
+  }
+
+  /// Called by [setPixels] to report overscroll when an attempt is made to
+  /// change the [pixels] position. Overscroll is the amount of change that was
+  /// not applied to the [pixels] value.
+  void didOverscrollBy(double value) {
+    assert(activity.isScrolling);
+    activity.dispatchOverscrollNotification(cloneMetrics(), context.notificationContext, value);
+  }
+
+  /// Called by [updateUserScrollDirection] to report that the
```

Comment:
This doc doesn't make sense, since updateUserScrollDirection doesn't exist on this class.
Should we move this function to the subclass?